mergify_engine/initial_configuration.py

```diff
@@ -0,0 +1,111 @@
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import logging
+import yaml
+
+import github
+
+LOG = logging.getLogger(__name__)
+
+INITIAL_CONFIG_BRANCH = "mergify/initial-config"
+
+
+def sanitize_context(context):
+    # NOTE(sileht): Some CI report different for pr and branch
+    # We can't require branch testing on PR, so remove the detail
+    # * travis: remove /pr, /push
+    if context.endswith("/pr"):
+        context = context[:-3]
+    elif context.endswith("/push"):
+        context = context[:-5]
+    return context
+
+
+def create_pull_request(installation_token, repo):
+    pulls = list(repo.get_pulls(state="all"))[:5]
+    contexts = set()
+    for pull in pulls:
+        contexts.update(
+            sanitize_context(s.context)
+            for s in repo.get_commit(pull.head.sha).get_statuses()
+            if not s.context.startswith("mergify")
+        )
+
+    mergify_config = {
+        "rules": {
+            "default": {
+                "protection": {
+                    "required_status_checks": {
+                        "strict": True,
+                        "contexts": list(contexts),
+                    },
+                    "required_pull_request_reviews": {
+                        "required_approving_review_count": 1
+                    },
+                },
+                "merge_strategy": {
+                    "method": "rebase"
+                }
+            }
+        }
+    }
+
+    content = yaml.dump(mergify_config, default_flow_style=False)
+
+    default_branch = repo.get_branch(repo.default_branch)
+
+    message = "Initial configuration of Mergify"
+    parents = [repo.get_git_commit(default_branch.commit.sha)]
+    tree = repo.create_git_tree([
+        github.InputGitTreeElement(".mergify.yml", "100644", "blob", content)
+    ], base_tree=default_branch.commit.commit.tree)
+    commit = repo.create_git_commit(message, tree, parents)
+    repo.create_git_ref("refs/heads/%s" % INITIAL_CONFIG_BRANCH, commit.sha)
+    repo.create_pull(
+        title="Initial configuration of Mergify",
+        body="""This is an initial configuration of Mergify.
+
+In this configuration proposal, enable the the [strict workflow](https://doc.mergify.io/strict-workflow.html), required one reviewer approval, use `rebase` as [merge strategy](https://doc.mergify.io/configuration.html#merge-strategy). And the required [status checks](https://doc.mergify.io/configuration.html#required-status-checks) have been discovered from the 5 more recents pull request of your project:
```

Comment:
This pull request enables the [strict workflow](https://doc.mergify.io/strict-workflow.html), requires one reviewer approval, uses `rebase` as the [merge strategy](https://doc.mergify.io/configuration.html#merge-strategy).

Those required [status checks](https://doc.mergify.io/configuration.html#required-status-checks) have been discovered from the 5 more recent pull requests of your project: